File: TemporaryMqLibrary/src/main/java/cattt.temporary.mq/wrapper/MqServiceConnection.java
package cattt.temporary.mq.wrapper;

import android.content.ComponentName;
import android.os.IBinder;

import cattt.temporary.mq.base.MqBinder;
import cattt.temporary.mq.base.MqService;
import cattt.temporary.mq.base.ServiceConnectionException;
import cattt.temporary.mq.callback.OnConnectionListener;
import cattt.temporary.mq.logger.Log;

public class MqServiceConnection extends BaseServiceConnection {
    private static Log logger = Log.getLogger(MqServiceConnection.class);

    private OnConnectionListener mListener;

    public MqServiceConnection addOnConnectionListener(OnConnectionListener listener) {
        if (listener == null) {
            new NullPointerException("param cannot be null.");
        }
        this.mListener = listener;
        return this;
    }


    protected MqServiceConnection() {
    }

    @Override
    protected String getConnectionType() {
        return MqServiceConnection.class.getSimpleName();
    }

    @Override
    protected String getImplicitAction() {
        return MqService.ACTION;
    }

    @Override
    protected String getOwnCategory() {
        return MqService.CATEGORY;
    }

    private void startConnect() throws ServiceConnectionException {
        if (getBinder() == null) {
            throw new ServiceConnectionException("Service is not connected");
        }
        ((MqBinder) getBinder()).startConnect();
    }

    public boolean isConnected(){
        if (getBinder() == null) {
            return false;
        }
        return ((MqBinder) getBinder()).isConnected();
    }

    public void publishMessage(String topic, String message) throws ServiceConnectionException {
        if (getBinder() == null) {
            throw new ServiceConnectionException("Service is not connected");
        }
        ((MqBinder) getBinder()).publishMessage(topic, message);
    }

    public void subscribe(String[] topic) throws ServiceConnectionException {
        if (getBinder() == null) {
            throw new ServiceConnectionException("Service is not connected");
        }
        ((MqBinder) getBinder()).subscribe(topic);
    }

    @Override
    public void onServiceConnected(ComponentName name, IBinder service) {
        super.onServiceConnected(name, service);
        startConnect();
        if (mListener != null) {
            mListener.onConnected();
        }
    }

    @Override
    public void onServiceDisconnected(ComponentName name) {
        super.onServiceDisconnected(name);
        if (mListener != null) {
            mListener.onDisconnected();
        }
    }

    private static final class Helper {
        private static final MqServiceConnection INSTANCE = new MqServiceConnection();
    }

    public static MqServiceConnection get() {
        return Helper.INSTANCE;
    }
}

File: CattMqKtLibs/src/main/java/catt.kt.libs.mq/core/MqServiceConnection.kt
package catt.kt.libs.mq.core

import android.content.ComponentName
import android.os.IBinder
import catt.kt.libs.mq.listeners.OnServiceConnectionListener
import catt.kt.libs.mq.wrapper.ServiceConnectionImpl
import java.lang.ref.WeakReference

internal class MqServiceConnection(
    connectionType: String = MqServiceConnection::class.java.simpleName,
    implicitAction: String = "catt.mq.libs.ACTION_CONNECTION_MQTT",
    ownCategory: String = "k632ioRVbA7xT8sf9RjGlH0v6nBthjmYKQNT49HvjJ0KPPltJRI51tNjdsHzh6nJ"
) : ServiceConnectionImpl(connectionType, implicitAction, ownCategory) {

    private var listenerWeakReference: WeakReference<OnServiceConnectionListener>? = null


    private fun connect() {
        binder ?: throw IllegalArgumentException("Service is not connected")
        (binder as MqBinder).connect()
    }

    val isConnected: Boolean
        get() {
            binder ?: return false
            return (binder as MqBinder).isConnected
        }

    fun publishMessage(topic: String, message: String) {
        binder ?: throw IllegalArgumentException("Service is not connected")
        (binder as MqBinder).publishMessage(topic, message)
    }

    fun subscribe(topic: Array<String>) {
        binder ?: throw IllegalArgumentException("Service is not connected")
        (binder as MqBinder).subscribe(topic)
    }

    fun setOnServiceConnectionListener(listener: OnServiceConnectionListener): ServiceConnectionImpl {
        listenerWeakReference = WeakReference(listener)
        return this@MqServiceConnection
    }

    override fun onServiceConnected(name: ComponentName, service: IBinder) {
        super.onServiceConnected(name, service)
        if (!isConnected) connect()
        listenerWeakReference?.get()?.onServiceConnected(name)
    }

    override fun onServiceDisconnected(name: ComponentName) {
        super.onServiceDisconnected(name)
        listenerWeakReference?.get()?.onServiceDisconnected(name)
    }

    companion object {
        private object Helper {
            val INSTANCE: MqServiceConnection by lazy { MqServiceConnection() }
        }

        @JvmStatic
        fun get(): MqServiceConnection {
            return Helper.INSTANCE
        }
    }
}